proposed/0000-UDP-hole-punching.md

```diff
@@ -0,0 +1,192 @@
+- Feature Name: UDP-hole-punching
+- Type new: feature
+- Related components: crust
+- Start Date: 22-09-2015
+- RFC PR: (leave this empty)
+- Issue number: (leave this empty)
+
+# Summary
+
+In this text we describe an implementation of UDP hole punching process
+to allow for P2P connections of nodes behind NATs in a fully decentralised
+network.
+
+# Motivation
+
+Most users of the Crust library are expected to run the software on home PCs connected
+to the internet through a router. Routers usually implement some kind of a firewall to
+prevent anyone from the outside internet to connect to user's PC directly. As Crust
+is a library that concentrates at P2P connections, it should try available techniques
+to circumvent these firewall restrictions and UDP hole punching is one such
+technique.
+
+Additionally, traditional UDP hole punching process involves a well known
+server to tell each peer's their external UDP endpoints. Since our intent
+is a fully decentralised system, this text also describes how nodes to which
+we're already connected can be used as a replacement for such server.
+
+In the following text the reader is expected to know basic NAT types and how they work.
+The section 'Methods of translation' on [wikipedia](https://en.wikipedia.org/wiki/Network_address_translation)
+gives a nice overview.
+
+# Detailed design
+
+The UDP hole punching process described here consists of two steps:
+
+1. Finding an external mapping of a UDP socket
+2. Do the actual hole punching.
+
+## Finding an external mapping of a UDP socket
+
+Suppose, we have two nodes `A` and `B`, where each one is possibly behind a NAT and
+they want to connect to each other.  Routing needs to send `A` an endpoint of `B` and
+vice versa. But `A` and `B` don’t know their public
+endpoints a priori (they may know their IP addresses from previous connection establishments,
+but not their ports). Each node therefore needs to create a UDP socket and use it to ask some
+other node it is connected to what the public endpoints of the UDP socket is. We’ll refer
+to such helping nodes `C(A)` and `C(B)` (it needs not be the same node for `A` and `B`).
+We'll call `U(C(X))` the UDP port on which X can contact `C(X)` in order to find its public
+endpoint. `X` finds out about `U(C(X))` through the initial handhsake, that is, when
+`X` connects to `C(X)`.
```

Comment:
If the scenario where no more than one network card is possible, then the use of socket address instead socket port is not more robust in any way.

But shouldn't we allow this extra bit of robustness, even if very few users will experience it? Also, the added complexity is very little for this change.